content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,323 @@
+
+```{r}
+out <- modify_in(x, list("foo", 2), as.character)
+str(out)
+```
+
+
+### Stricter pluck()
+
+Thanks to Daniel Barnett (@daniel-barnett on Github), `pluck()` now has a stricter cousin `chuck()`. Whereas `pluck()` is very permissive regarding non-existing locations and returns `NULL` in these cases, and `[[` inconsistently returns `NULL`, `NA`, or throws an error, `chuck()` fails consistently with informative messages (i.e., it "chucks" an error message):
+
+```{r, error = TRUE}
+pluck(list(1), "foo")
+
+chuck(list(1), "foo")
+```
+
+
+## Rates
+
+Thanks to Richie Cotton (@richierocks) and Ian Lyttle (@ijlyttle), purrr gains a function operator to make a function call itself repeatedly when an error occurs.
+
+```{r, error = TRUE}
+counter <- 0
+
+f <- function(...) {
+  if (counter < 2) {
+    counter <<- counter + 1
+    stop("tilt!")
+  }
+  "result"
+}
+
+f()
+```
+
+If the function is wrapped with `insistently()`, it will try a few times before giving up:
+
+```{r}
+# Reset counter
+counter <- 0
+
+f2 <- insistently(f)
+f2()
+```
+
+Another rate limiting function is `slowly()`. While `insistently()` loops by itself, `slowly()` is designed to be used in your own loops, for instance in a map iteration:
+
+```{r}
+f <- function(...) print(Sys.time())
+
+walk(1:3, f)
+
+walk(1:3, slowly(f))
+```
+
+`slowly()` uses a constant rate by default while `insistently()` uses a backoff rate. The rate limiting can be configured with optional jitter via `rate_backoff()` and `rate_delay()`, which implement exponential backoff rate and constant rate respectively.
+
+```{r}
+walk(1:3, slowly(f, rate_backoff(max_times = Inf)))
+```
+
+## Map and reduce improvements
+
+### `map_if()`... or else?
+
+If you like using `map_if()`, perhaps you'll find the new `.else` argument useful. `.else` is a function applied to elements for which the predicate is `FALSE`:
+
+```{r}
+map_if(iris, is.numeric, mean, .else = nlevels)
+```
+
+
+### New `map_at()` features
+
+Colin Fay (@ColinFay) has added support for tidyselect expressions to `map_at()` and other `_at` mappers. This brings the interface of these functions closer to scoped functions from the dplyr package, such as `dplyr::mutate_at()`. Note that `vars()` is currently not reexported from purrr, so you need to use `dplyr::vars()` or `ggplot2::vars()` for the time being.
+
+```{r}
+suppressMessages(library("dplyr"))
+
+x <- list(
+  foo = 1:5,
+  bar = 6:10,
+  baz = 11:15
+)
+
+map_at(x, vars(starts_with("b")), mean)
+```
+
+`map_at()` now also supports negative selections:
+
+```{r}
+map_at(x, -2, `*`, 1000)
+```
+
+
+### Early termination of reduction
+
+‘reduce()’ is an operation that combines the elements of a vector into a single value by calling a binary function repeatedly with the result so far and the next input of a vector. `reduce()` and its variant `accumulate()` now support early termination of the reduction. To halt the computation, just return the last value wrapped in a `done()` box:
```

Comment:
```suggestion
`reduce()` is an operation that combines the elements of a vector into a single value by calling a binary function repeatedly with the result so far and the next input of a vector. `reduce()` and its variant `accumulate()` now support early termination of the reduction. To halt the computation, just return the last value wrapped in a `done()` box:
```

Or maybe you really are trying to talk about "reduce" as a concept and not the function?